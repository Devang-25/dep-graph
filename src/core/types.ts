// using .ts instead of .d.ts for this file,
//   as otherwise it's referenced from the auto-generated `index.d.ts`,
//   but not copied over to the `./dist` folder
//   and thus fails the compilation of a typescript project that imports this lib

export interface Pkg {
  name: string;
  version?: string;
}

export interface PkgInfo {
  name: string;
  version?: string;
  // NOTE: consider adding in the future
  // requires?: {
  //   name: string;
  //   versionRange: string;
  //   dev?: boolean;
  //   exclusions?: string[];
  // }[];
  // resolved?: {
  //   registry: string;
  //   url: string;
  // };
  // parentManifestUri?: string;
  // issues?: {
  //   id: string;
  //   type: string;
  //   title: string;
  //   severity: string;
  // }[];
}

export interface VersionProvenance {
  type: string;
  location: string;
  property?: {
    name: string;
  };
}

export interface NodeInfo {
  versionProvenance?: VersionProvenance;
  labels?: {
    [key: string]: string;
  };
}

export interface GraphNode {
  nodeId: string;
  pkgId: string;
  info?: NodeInfo;
  deps: Array<{
    nodeId: string;
    // NOTE: consider adding later:
    // meta?: JsonMap;
  }>;
}

export interface PkgManager {
  name: string;
  version?: string;
  repositories?: Array<{
    alias: string;
  }>;
}

export interface DepGraphData {
  schemaVersion: string;
  pkgManager: PkgManager;
  pkgs: Array<{
    id: string;
    info: PkgInfo;
  }>;
  graph: {
    rootNodeId: string;
    nodes: GraphNode[];
  };
}

export interface DepGraph {
  readonly pkgManager: PkgManager;
  readonly rootPkg: PkgInfo;
  getPkgs(): PkgInfo[];
  toJSON(): DepGraphData;
  pkgPathsToRoot(pkg: Pkg): PkgInfo[][];
  countPathsToRoot(pkg: Pkg): number;
}

// NOTE/TODO(shaun): deferring any/all design decisions here
// Revisit when we actually start using things
export interface DepGraphInternal extends DepGraph {
  readonly rootNodeId: string;
  getNode(nodeId: string): NodeInfo;
  getNodePkg(nodeId: string): PkgInfo;
  getPkgNodeIds(pkg: Pkg): string[];
  getNodeDepsNodeIds(nodeId: string): string[];
  getNodeParentsNodeIds(nodeId: string): string[];
  hasCycles(): boolean;
}
