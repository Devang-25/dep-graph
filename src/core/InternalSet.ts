export class InternalSet {
  state = {};

  constructor(oldState?: InternalSet) {
    Object.keys(oldState?.state || {}).forEach(k=>this.state[k] = true)
    return this;
  }

  add(k) {
    this.state[k] = true;
    return this;
  }

  has(k: string): boolean {
    return this.state[k];
  }
}
