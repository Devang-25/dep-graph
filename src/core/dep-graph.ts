import * as _ from 'lodash';
import * as graphlib from 'graphlib';
import * as types from './types';

export {
  DepGraphImpl,
};

class DepGraphImpl implements types.DepGraphInternal {
  public static SCHEMA_VERSION = '2.0.0';

  private _graphInstance: graphlib.Graph;
  private _pkgValues: types.PkgInfo[];
  private _pkgNodeIds: { [pkgId: string]: Set<string> };

  public constructor(private data: types.DepGraphData) {
    this._pkgValues = _.values(data.pkgs);
  }

  get pkgManager() {
    return this.data.pkgManager;
  }

  get rootPkg(): types.PkgInfo {
    return this.data.pkgs.root;
  }

  public getPkgs(): types.PkgInfo[] {
    return this._pkgValues;
  }

  public getNodePkg(nodeId: string): types.PkgInfo {
    const node = this.data.graph[nodeId];
    if (!node) {
      throw new Error(`no such node: ${nodeId}`);
    }
    return this.data.pkgs[node.pkgId];
  }

  public getPkgNodeIds(pkg: types.Pkg): string[] {
    if (pkg === this.rootPkg ||
      (pkg.name === this.rootPkg.name && pkg.version === this.rootPkg.version)) {
      return Array.from(this.pkgNodeIds.root);
    }

    const pkgId = `${pkg.name}@${pkg.version || ''}`;
    if (!this.data.pkgs[pkgId]) {
      throw new Error(`no such pkg: ${pkgId}`);
    }

    return Array.from(this.pkgNodeIds[pkgId]);
  }

  public getNodeDepsNodeIds(nodeId: string): string[] {
    const deps = this.graphInstance.successors(nodeId);
    if (!deps) {
      throw new Error(`no such node: ${nodeId}`);
    }
    return deps;
  }

  public getNodeParentsNodeIds(nodeId: string): string[] {
    const parents = this.graphInstance.predecessors(nodeId);
    if (!parents) {
      throw new Error(`no such node: ${nodeId}`);
    }
    return parents;
  }

  public hasCycles(): boolean {
    return !graphlib.alg.isAcyclic(this.graphInstance);
  }

  public pkgPathsToRoot(pkg: types.Pkg): types.PkgInfo[][] {
    // TODO: implement cycles support
    if (this.hasCycles()) {
      throw new Error('pkgPathsToRoot does not support cyclic graphs yet');
    }

    const pathsToRoot: types.PkgInfo[][] = [];

    const nodeIds = this.getPkgNodeIds(pkg);
    if (nodeIds) {
      for (const id of nodeIds) {
        pathsToRoot.push(...this.pathsFromNodeToRoot(id));
      }
    }
    // note: sorting to get shorter paths first -
    //  it's nicer - and better resembles older behaviour
    return pathsToRoot.sort((a, b) => a.length - b.length);
  }

  public toJSON(): types.DepGraphData {
    return this.data;
  }

  private pathsFromNodeToRoot(nodeId: string): types.PkgInfo[][] {
    const parentNodesIds = this.getNodeParentsNodeIds(nodeId);
    if (parentNodesIds.length === 0) {
      return [[this.getNodePkg(nodeId)]];
    }
    const allPaths: types.PkgInfo[][] = [];
    parentNodesIds.map((id) => {
      const out = this.pathsFromNodeToRoot(id).map((path) => {
        return [this.getNodePkg(nodeId)].concat(path);
      });
      allPaths.push(...out);
    });
    return allPaths;
  }

  private get pkgNodeIds(): { [pkgId: string]: Set<string> } {
    if (this._pkgNodeIds) {
      return this._pkgNodeIds;
    }
    const pkgNodes: { [pkgId: string]: Set<string> } = {};
    for (const nodeId of Object.keys(this.data.graph)) {
      const node = this.data.graph[nodeId];
      const pkgId = node.pkgId;
      if (!pkgNodes[pkgId]) {
        pkgNodes[pkgId] = new Set();
      }
      pkgNodes[pkgId].add(nodeId);
    }
    this._pkgNodeIds = pkgNodes;
    return pkgNodes;
  }

  private get graphInstance(): graphlib.Graph {
    if (this._graphInstance) {
      return this._graphInstance;
    }
    const graph = new graphlib.Graph({
      directed: true,
      multigraph: false,
      compound: false,
    });
    // const pkgs = {};

    // for (const pkgId of Object.keys(this.data.pkgs)) {
    //   const pkg = this.data.pkgs[pkgId];
    //   pkgs[pkgId] = pkg.version ? pkg : { ...pkg, version: null };
    // }

    for (const nodeId of Object.keys(this.data.graph)) {
      const node = this.data.graph[nodeId];
      graph.setNode(nodeId, { pkgId: node.pkgId });
    }

    for (const nodeId of Object.keys(this.data.graph)) {
      const node = this.data.graph[nodeId];
      for (const depNodeId of node.deps) {
        graph.setEdge(nodeId, depNodeId.nodeId);
      }
    }

    this._graphInstance = graph;
    return graph;
  }
}
