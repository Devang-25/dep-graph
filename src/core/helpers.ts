export type Index = Record<string, Set<string>>;

export function addToIndex(index: Index, key: string, value: string)  {
  if (!index[key]) {
    index[key] = new Set();
  }
  index[key].add(value);
}
