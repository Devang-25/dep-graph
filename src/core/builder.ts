import * as graphlib from 'graphlib';
import * as types from './types';
import { DepGraphImpl } from './dep-graph';

export {
  DepGraphBuilder,
};

class DepGraphBuilder {
  private _pkgs: {
    root: types.PkgInfo;
    [pkgId: string]: types.PkgInfo;
  };
  private _pkgNodes: { [pkgId: string]: Set<string> } = {};

  private _graph: graphlib.Graph;
  private _pkgManager: types.PkgManager;

  public constructor(pkgManager: types.PkgManager, rootPkg?: types.PkgInfo) {
    const graph = new graphlib.Graph({
      directed: true,
      multigraph: false,
      compound: false,
    });

    this._pkgs = {
      root: rootPkg || {
        name: '_root',
        version: '0.0.0',
      },
    };

    graph.setNode('root', { pkgId: 'root' });
    this._pkgNodes.root = new Set(['root']);
    this._graph = graph;
    this._pkgManager = pkgManager;
  }

  // TODO: this can create disconnected nodes
  public addPkgNode(pkgInfo: types.PkgInfo, nodeId: string) {
    if (nodeId === 'root') {
      throw new Error('DepGraphBuilder.addPkgNode() cant override root node');
    }

    const pkgId = `${pkgInfo.name}@${pkgInfo.version || ''}`;

    this._pkgs[pkgId] = pkgInfo;
    this._pkgNodes[pkgId] = this._pkgNodes[pkgId] || new Set();
    this._pkgNodes[pkgId].add(nodeId);

    this._graph.setNode(nodeId, { pkgId });
  }

  // TODO: this can create cycles
  public connectDep(parentNodeId: string, depNodeId: string) {
    if (!this._graph.hasNode(parentNodeId)) {
      throw new Error('parentNodeId does not exist');
    }

    if (!this._graph.hasNode(depNodeId)) {
      throw new Error('depNodeId does not exist');
    }

    this._graph.setEdge(parentNodeId, depNodeId);
  }

  public build(): types.DepGraph {
    const nodeIds = this._graph.nodes();

    const nodes: any = nodeIds.reduce((acc, nodeId: string) => {
      const deps = (this._graph.successors(nodeId) || [])
        .map((depNodeId) => ({ nodeId: depNodeId }));

      acc[nodeId] = {
        pkgId: this._graph.node(nodeId).pkgId,
        deps,
      };
      return acc;
    }, {});

    return new DepGraphImpl({
      schemaVersion: '2.0.0',
      pkgManager: this._pkgManager,
      pkgs: this._pkgs,
      graph: nodes,
    });
  }
}
