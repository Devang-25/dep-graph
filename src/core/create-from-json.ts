import * as _ from 'lodash';
import * as semver from 'semver';

import { DepGraph, DepGraphData } from './types';
import { DepGraphImpl } from './dep-graph';

export const SUPPORTED_SCHEMA_RANGE = '^2.0.0';

export function createFromJSON(depGraphData: DepGraphData): DepGraph {
  validateDepGraphData(depGraphData);
  return new DepGraphImpl(depGraphData);
}

function assert(condition: any, msg: string) {
  if (!condition) {
    throw new Error(msg);
  }
}

function validateDepGraphData(data: DepGraphData) {
  assert(data.schemaVersion, `schemaVersion missing`);
  assert(semver.satisfies(data.schemaVersion, SUPPORTED_SCHEMA_RANGE),
    `schemaVersion not in "${SUPPORTED_SCHEMA_RANGE}"`);
  assert(data.pkgManager && data.pkgs && data.graph,
    'bad data format');

  assert(data.pkgManager.name, '.pkgManager.name is missing');
  assert(data.pkgs.root, `root pkg missing`);
  assert(data.pkgs.root.name, `root pkg missing`);
  assert(data.graph.root, `root graph node is missing`);
  assert(data.graph.root.pkgId === 'root',
    `the root node .pkgId must be "root", but got ${data.graph.root.pkgId}`);

  // validate packages
  for (const pkgId of Object.keys(data.pkgs)) {
    const pkg = data.pkgs[pkgId];
    assert(!!pkg, 'empty pkg');
    assert(pkg.name, 'some .pkgs elements have no .name field');
    // NOTE: this name@version check is very strict,
    // we can relax it later, it just makes things easier now
    assert(pkgId === 'root' || pkgId === `${pkg.name}@${pkg.version || ''}`,
      'non-root pkg id must be name@version');
  }

  // validate nodes
  const pkgIdsSeen: Set<string> = new Set();
  for (const [nodeId, node] of _.entries(data.graph)) {
    pkgIdsSeen.add(node.pkgId);
    assert(node.pkgId !== 'root' || nodeId === 'root',
      'root pkg should have exactly one instance node');
    assert(data.pkgs[node.pkgId], 'node points to a non-existing pkgId');
    for (const depNodeId of node.deps) {
      // TODO(shaun): better error
      assert(depNodeId !== 'root', `"root" is not really the root`);
      // TODO(shaun): test
      assert(!data.graph[depNodeId], 'node depends on a non-existing nodeId');
    }
  }
  assert(pkgIdsSeen.size === Object.keys(data.pkgs).length,
    'not all pkgs have instance nodes');

  function nodesReachableFrom(
    nodeId: string,
    reachable: Set<string> = new Set()) {

    reachable.add(nodeId);
    const node = data.graph[nodeId];
    for (const dep of node.deps) {
      if (!reachable.has(dep.nodeId)) {
        nodesReachableFrom(dep.nodeId, reachable);
      }
    }
    return reachable;
  }

  assert(nodesReachableFrom('root').size === Object.keys(data.graph).length,
    'not all graph nodes are reachable from root');
}
