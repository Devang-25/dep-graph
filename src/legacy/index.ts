import * as _ from 'lodash';
import * as crypto from 'crypto';
import * as types from '../core/types';
import { DepGraphBuilder } from '../core/builder';

export {
  depTreeToGraph,
  DepTree,
};

interface DepTreeDep {
  name?: string; // shouldn't, but might happen
  version?: string; // shouldn't, but might happen
  dependencies?: {
    [depName: string]: DepTreeDep,
  };
}

interface DepTree extends DepTreeDep {
  targetOS?: {
    name: string;
    version: string;
  };
}

async function depTreeToGraph(depTree: DepTree, pkgManagerName: string): Promise<types.DepGraph> {
  const rootPkg = {
    name: depTree.name,
    version: depTree.version,
  };

  const pkgManagerInfo: types.PkgManager = {
    name: pkgManagerName,
  };

  const targetOS = depTree.targetOS;
  if (targetOS) {
    pkgManagerInfo.repositories = [
      {
        alias: `${targetOS.name}:${targetOS.version}`,
      },
    ];
  }

  const builder = new DepGraphBuilder(pkgManagerInfo, rootPkg);

  await buildGraph(builder, depTree, depTree.name, true);

  const depGraph = await builder.build();

  return shortenNodeIds(depGraph as types.DepGraphInternal);
}

async function buildGraph(
  builder: DepGraphBuilder, depTree: DepTreeDep, pkgName: string, isRoot = false): Promise<string> {

  const getNodeId = (name, version, hashId) => `${name}@${version || ''}|${hashId}`;

  const depNodesIds = [];

  const hash = crypto.createHash('sha1');

  const deps = depTree.dependencies || {};
  const depNames = _.keys(deps).sort();
  for (const depName of depNames) {
    const dep = deps[depName];

    const subtreeHash = await buildGraph(builder, dep, depName);

    const depPkg = {
      name: depName,
      version: dep.version,
    };

    const depNodeId = getNodeId(depPkg.name, depPkg.version, subtreeHash);

    depNodesIds.push(depNodeId);

    builder.addPkgNode(depPkg, depNodeId);

    hash.update(depNodeId);
  }

  const treeHash = depNames.length ? hash.digest('hex') : 'leaf';

  let pkgNodeId;
  if (isRoot) {
    pkgNodeId = builder.rootNodeId;
  } else {
    // we don't assume depTree has a .name to support output of `npm list --json`
    const pkg = {
      name: pkgName,
      version: depTree.version,
    };
    pkgNodeId = getNodeId(pkg.name, pkg.version, treeHash);
    builder.addPkgNode(pkg, pkgNodeId);
  }

  for (const depNodeId of depNodesIds) {
    builder.connectDep(pkgNodeId, depNodeId);
  }

  if (depNodesIds.length > 0) {
    await spinTheEventLoop();
  }
  return treeHash;
}

async function shortenNodeIds(depGraph: types.DepGraphInternal): Promise<types.DepGraph> {
  const builder = new DepGraphBuilder(depGraph.pkgManager, depGraph.rootPkg);

  const nodesMap: {[key: string]: string} = {};

  // create nodes with shorter ids
  for (const pkg of depGraph.getPkgs()) {
    const nodeIds = depGraph.getPkgNodeIds(pkg);
    for (let i = 0; i < nodeIds.length; i++) {
      const nodeId = nodeIds[i];
      if (nodeId === depGraph.rootNodeId) {
        continue;
      }

      let newNodeId: string;
      if (nodeIds.length === 1) {
        newNodeId = `${trimAfterLastSep(nodeId, '|')}`;
      } else {
        newNodeId = `${trimAfterLastSep(nodeId, '|')}|${i + 1}`;
      }

      nodesMap[nodeId] = newNodeId;
      builder.addPkgNode(pkg, newNodeId);
    }

    await spinTheEventLoop();
  }

  // connect nodes
  for (const pkg of depGraph.getPkgs()) {
    for (const nodeId of depGraph.getPkgNodeIds(pkg)) {
      for (const depNodeId of depGraph.getNodeDepsNodeIds(nodeId)) {
        const parentNode = nodesMap[nodeId] || nodeId;
        const childNode = nodesMap[depNodeId] || depNodeId;

        builder.connectDep(parentNode, childNode);
      }
    }

    await spinTheEventLoop();
  }

  return builder.build();
}

async function spinTheEventLoop() {
  return new Promise((resolve) => setImmediate(resolve));
}

function trimAfterLastSep(str: string, sep: string) {
  return str.slice(0, str.lastIndexOf(sep));
}
