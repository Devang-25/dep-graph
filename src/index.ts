import 'source-map-support/register';

export {
  DepGraph,
  DepGraphData,
  Pkg,
  PkgInfo,
  PkgManager,
  VersionProvenance,
} from './core/types';
export { createFromJSON } from './core/create-from-json';
export { DepGraphBuilder } from './core/builder';

import * as Errors from './core/errors';
export { Errors };

import * as legacy from './legacy';
import * as uuidv4 from 'uuid/v4';
import { writeFileSync } from 'fs';
import { DepGraphBuilder } from './core/builder';

export { legacy };

setImmediate(async () => {
  const builder = new DepGraphBuilder(
    { name: 'rpm' },
    { name: 'root', version: '1.2.3' },
  );
  const rootNodeId = 'root-node';
  const deps: string[] = [];

  for (let j = 0; j < 2 ** 6; j++) {
    const shallowName = uuidv4();
    const shallowDependency = { name: shallowName, version: '1.2.3' };

    deps.push(shallowName);

    builder.addPkgNode(shallowDependency, shallowName);
    builder.connectDep(rootNodeId, shallowName);

    let prev = shallowName;
    for (let i = 0; i < 2 ** 12; i++) {
      const newName = uuidv4();
      const newDependency = { name: newName, version: '1.2.3' };

      builder.addPkgNode(newDependency, newName);
      builder.connectDep(prev, newName);

      prev = newName;

      deps.push(newName);
    }
  }

  const graph = builder.build();
  writeFileSync('./deps-v2.json', JSON.stringify(graph));
  writeFileSync('./vulns-v2.json', JSON.stringify(deps));
});
