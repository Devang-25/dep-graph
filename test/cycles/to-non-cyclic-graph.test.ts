import * as depGraphLib from '../../src';
import * as helpers from '../helpers';
import { mapToNonCyclicGraph } from '../../src/legacy';
import { DepGraphInternal } from '../../src/core/types';
import { DepGraph, DepGraphData } from '../../src';

describe('mapToNonCyclicGraph', () => {
  describe('cyclic graph', () => {
    let cyclicDepGraphData: DepGraphData;
    let cyclicDepGraph: DepGraph;
    let nonCyclicDepGraph: DepGraphInternal;

    beforeEach(async () => {
      cyclicDepGraphData = helpers.loadFixture('cyclic-dep-graph.json');
      cyclicDepGraph = depGraphLib.createFromJSON(cyclicDepGraphData);

      nonCyclicDepGraph = (await depGraphLib.legacy.mapToNonCyclicGraph(
        cyclicDepGraph,
      )) as DepGraphInternal;
    });

    test('should not change original depGraph', async () => {
      const initialDepGraphData = helpers.loadFixture('cyclic-dep-graph.json');

      expect(cyclicDepGraph.toJSON()).toEqual(initialDepGraphData);
    });

    test('should return a non-cyclic graph', async () => {
      expect(nonCyclicDepGraph.hasCycles()).toBeFalsy();
    });

    test('should return a cyclic graph with "more or equal" nodes number', async () => {
      expect(nonCyclicDepGraph.getDepPkgs().length).toBeGreaterThanOrEqual(
        cyclicDepGraph.getDepPkgs().length,
      );
    });

    test('should return a non-cyclic graph - snapshot', async () => {
      expect(nonCyclicDepGraph).toMatchSnapshot();
    });
  });

  test('should return the same graph for non-cyclic graph', async () => {
    const nonCyclicDepGraphData = helpers.loadFixture('goof-graph.json');
    const cyclicDepGraph = depGraphLib.createFromJSON(nonCyclicDepGraphData);

    const resultDepGraph = (await depGraphLib.legacy.mapToNonCyclicGraph(
      cyclicDepGraph,
    )) as DepGraphInternal;
    expect(resultDepGraph).toStrictEqual(cyclicDepGraph);
  });
});
