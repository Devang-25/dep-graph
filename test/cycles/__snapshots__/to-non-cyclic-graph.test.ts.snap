// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToNonCyclicGraph cyclic graph should return a non-cyclic graph - snapshot 1`] = `
Object {
  "graph": Object {
    "nodes": Array [
      Object {
        "deps": Array [
          Object {
            "nodeId": "foo@2|x",
          },
        ],
        "nodeId": "toor",
        "pkgId": "toor@1.0.0",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "bar@3|x",
          },
        ],
        "nodeId": "foo@2|x",
        "pkgId": "foo@2",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "baz@4|x",
          },
        ],
        "nodeId": "bar@3|x",
        "pkgId": "bar@3",
      },
      Object {
        "deps": Array [
          Object {
            "nodeId": "foo@2|x_pruned",
          },
        ],
        "nodeId": "baz@4|x",
        "pkgId": "baz@4",
      },
      Object {
        "deps": Array [],
        "info": Object {
          "labels": Object {
            "pruned": "cyclic",
          },
        },
        "nodeId": "foo@2|x_pruned",
        "pkgId": "foo@2",
      },
    ],
    "rootNodeId": "toor",
  },
  "pkgManager": Object {
    "name": "pip",
  },
  "pkgs": Array [
    Object {
      "id": "toor@1.0.0",
      "info": Object {
        "name": "toor",
        "version": "1.0.0",
      },
    },
    Object {
      "id": "foo@2",
      "info": Object {
        "name": "foo",
        "version": "2",
      },
    },
    Object {
      "id": "bar@3",
      "info": Object {
        "name": "bar",
        "version": "3",
      },
    },
    Object {
      "id": "baz@4",
      "info": Object {
        "name": "baz",
        "version": "4",
      },
    },
  ],
  "schemaVersion": "1.2.0",
}
`;
